### PR TITLE
ARROW-9082:  [Rust] - Stream reader fail when steam not ended with (opt…

### PR DESCRIPTION
**Pull request by eyalleshem**
…ional) 0xFFFFFFFF 0x00000000

### CODE REVIEW

File: rust/arrow/src/ipc/reader.rs [hunk 1]
```diff
@@ -793,7 +793,21 @@ impl<R: Read> StreamReader<R> {
...
+            Ok(()) => (),
+            Err(e) => {
+                if e.kind() == std::io::ErrorKind::UnexpectedEof {
+                    // Handle EOF without the "0xFFFFFFFF 0x00000000"
+                    // valid according to:
+                    // https://arrow.apache.org/docs/format/Columnar.html#ipc-streaming-format
+                    self.finished = true;
+                    return Ok(None);
+                }
+                return Err(ArrowError::from(e));
```
Comment (andygrove) [inline rust/arrow/src/ipc/reader.rs:807]: nit: could you remove the `return` statements?

perhaps:

```rust
if e.kind() == std::io::ErrorKind::UnexpectedEof {
  ...
  Ok(None)
} else {
  Err(ArrowError::from(e))
}
```

Another option:

```rust
Err(e) if e.kind() == std::io::ErrorKind::UnexpectedEof => {
  ...
  Ok(None)
}
Err(e) => Err(ArrowError::from(e))
```

Comment (eyalleshem) [reply rust/arrow/src/ipc/reader.rs:807]: I don't think that i could remove return  , As it's not the last statement of the function .  

for example i tried this : 
`if let Err(e) = self.reader.read_exact(&mut meta_size) {
            if e.kind() == std::io::ErrorKind::UnexpectedEof {
                // Handle EOF without the "0xFFFFFFFF 0x00000000"
                // valid according to:
                // https://arrow.apache.org/docs/format/Columnar.html#ipc-streaming-format
                self.finished = true;
                Ok(None)
            } else {
                Err(ArrowError::from(e))
            }
        } `

but the complier complain : 
`expected `()`, found enum `std::result::Result``

Comment (andygrove) [reply rust/arrow/src/ipc/reader.rs:807]: I see. This is returning from the whole function not the match .. ok, never mind then and thanks for looking at it.

**Review by andygrove — APPROVED** [review]
LGTM

File: rust/arrow/src/ipc/reader.rs (hunk 1 above)
Comment (andygrove) [inline rust/arrow/src/ipc/reader.rs:807]: One last option to consider:

```rust
return if e.kind() == std::io::ErrorKind::UnexpectedEof {
    ...
    Ok(None)
} else {
    Err(ArrowError::from(e))
}
```

Comment (eyalleshem) [reply rust/arrow/src/ipc/reader.rs]: done

**Review by nevi-me — APPROVED** [review]

**Pull request closed without merging**